### ISSUE TITLE
Compilation error if wildcard with `when` clause starts with an exclamation mark.

### ISSUE THREAD

**Issue by water-mizuu** [dart-lang/sdk#52220]
* Version: `Dart SDK version: 3.1.0-46.0.dev (dev) (Tue Apr 25 23:35:50 2023 -0700) on "windows_x64"`
* Windows 11

Basically, as stated in the title, the code does not compile if a switch-expression case has a wildcard and a `when` clause that starts with a negation.

```dart
import "dart:math";

final Random random = Random();
bool twoState() => switch (random.nextDouble()) {
      < 0.5 => false,
      >= 0.5 => true,
      _ => throw Error(),
    };

void main() {
  print(switch (twoState()) {
    _ when !twoState() => "hi", // Expected to find '=>'
    // Function expressions can't be named. Try removing the name, or moving the function expression to a function declaration statement.
    bool _ when !twoState() => "hi", // Works.
    true => 1,
    false => 0,
  });
}
```

**Comment by water-mizuu** [comment]
Note that this also fails when the expression is not a function call.

```dart
 print(switch (twoState()) {
    _ when !(true != false) => "hi",
    // Expected to find '=>'
    // 'true' can't be used as an identifier because it's a keyword. Try renaming this to be an identifier that isn't a keyword.
    // Expected to find ')'
    true => 1,
    false => 0,
  });
```

**Comment by mraleph** [comment]
tentatively marking as a parser issue. @jensjoha can you take a look?

**Comment by jensjoha** [comment]
I think this was fixed with 637dd76c7f926b7476bb35944bf7888e3d7d52f1.
